### ISSUE TITLE
Need a easy list of all the abstractions

### ISSUE THREAD

**Issue by isaacabraham** [mbraceproject/MBrace.Core#55]
I'm seeing more and more people having difficulties about getting up to speed with the abstractions of MBrace. Things like CloudVector have little documentation on it - the intellisense comments of each static method just refer to CloudVector but don't actually explain what it is. Is there a list of all the main abstractions e.g. on the website?

Going through the starter pack, I think there needs to be better explanations of what things are, what the use case of them is and how it fits in with the rest of the MBrace API overall. Perhaps there's a need for some easier exercises in the pack that use some dataset to work with rather than something abstract like prime numbers etc. etc.?

**Comment by dsyme** [comment]
Yes. First please feel free to modify any of the starter pack (though the learning sequence seems reasonable).

Let's also do a documentation bash ASAP, before the next release.

**Comment by dsyme** [comment]
Do you have any details of who/how/what of the people learning the technology?  Were they F# data scripters already?  Did they have a CPU or data problem they actually needed to solve, or did they just want to play?  Did they have a teacher or were going through by themselves (I think the latter is still very hard - I think we need some good introductory teaching-style videos). That sort of thing.

**Comment by isaacabraham** [comment]
One of the people was an MVP, an experienced C# developer with a reasonable amount of F# experience. The initial part I witnessed was them simply going through the starter packs - some parts are a bit too abstract - this individual just wanted to see the programming model but some of the examples are too complicated.

I think, looking at the starter pack, there needs to be some organisation around e.g. distributed compute, distributed data, messaging etc.. At the moment the pack jumps from one part to another to another.

The other issue is the API (which we all know will have some work done to it) - it's confusing to some people - you need to know *in advance* which arbitrary abstraction you need in order to achieve a given goal - and the namespaces and types are not predictable.

I'm thinking of either putting together some simple use cases and example solutions, and maybe some suggestions on changes to existing, or new abstractions.

**Comment by dsyme** [comment]
Which examples tripped them up?  

I've simplified the learning sequence here, please take a look: https://github.com/mbraceproject/MBrace.StarterKit/pull/13

**Comment by dsyme** [comment]
> The other issue is the API (which we all know will have some work done to it) - it's confusing to some people - you need to know in advance which arbitrary abstraction you need in order to achieve a given goal - and the namespaces and types are not predictable.

Any specifics?  The "cluster.$" completions?  The "MBrace.$" completions?  

BTW looking at the "MBrace.$" completions, it looks like all of that should go in a namespace "MBrace.Core".

**Comment by eiriktsarpalis** [comment]
I think that the best way to introduce non-obvious abstractions is demonstrating examples that necessitate them. Let me illustrate by example:

One of the major problems with mbrace as a programming model is that very subtle changes to code introduce enormous performance penalties in a hard-to-measure way. In other words, the nice sequential-looking code of cloud workflows tend to make users forget that this translates to closures, which may jump from machine to machine. Sometimes one needs to carry an acute understanding of how F# generates code under all the let bindings in order to reason about performance in mbrace.

For instance, it might be often assumed that since a variable exists in scope, it comes at no cost to reference it as if this was a shared memory computation: the code
```fsharp
cloud {
     let array = [|1 ... 1000000|]
     return! Cloud.Parallel [for i in 1 .. 200 -> cloud { return isPrime array.Length } ]
}
```
seems equivalent to
```fsharp
cloud {
     let array = [|1 ... 1000000|]
     let n = array.Length
     return! Cloud.Parallel [for i in 1 .. 200 -> cloud { return isPrime n } ]
}
```
But is it really? The first example is much slower than the second, because the parallel computation *captures the array in its closure*. This will cause it to be serialised and sent off 200 times. 

But what if the parallel computation is reliant on the whole dataset and not just the length? What do I do then? How do I minimise communication? The solution lies with CloudCell:
```fsharp
cloud {
    let! arrayRef = CloudCell.New([|1 .. 1000000000|], cacheByDefault = true)
    return! Cloud.Parallel [ for i in 1 .. 200 -> cloud { let! value = arrayRef.Value in return Array.sum value } ]
}
```
This will persist data exactly once to store - to be retrieved by the clusters on-demand. Given that I have enabled caching by default, this will happen about once per worker. Assuming that the cluster size is much lower than 200, I have eliminated excessive communication while stilling replicating the data across the cluster. At the same time, parallel closures do not reference the initial array, hence they do not capture it. 

One might be tempted to write it this way:
```fsharp
cloud {
    let! value = arrayRef.Value
    return! Cloud.Parallel [ for i in 1 .. 200 -> cloud { return Array.sum value } ]
}
```
however this invalidates the point of CloudCell; closures will capture its *payload* rather than the lightweight reference.

I think that incremental demonstrations like the above are best at explaining the purpose of every abstraction. It definitely beats the "here's the primitive, this is its API, here's what it does" approach. The initial naive examples should preferably demonstrate unexpected results rather than a vague claim of reduced performance, but this is not always achievable.

**Comment by isaacabraham** [comment]
@eiriktsarpalis: This is all true. How do we help people fall into the pit of success by doing the right thing first? Do we need higher level abstractions in order to restrict users from doing the "wrong" thing? Do the existing one need changing? Do we need warnings being emitted if the runtime detects abnormally large items being round tripped repeatedly? A good example of another system doing this is something like RavenDB, which will prevent deliberately prevent the user from querying the database and returning huge datasets unless they absolutely explicitly request to do so.

**Comment by isaacabraham** [comment]
@dsyme here's an example from earlier this week and one that I have fallen foul of as well in the past.

I have a file in blob storage that I want to read in as part of a distributed computation. I look through intellisense of MBrace.$ and see the following prospective candidates: -

1. CloudCacheable
2. CloudCell
3. CloudFile

CloudCell seems like it might be the right one, because it has a method on it called Cache. But then the file is already in blob, so I guess I need to use CloudFile. Why are there two different abstractions for dealing with data that gets persisted in blobs - one of them allows us to do read/ write operations and the other allows us to cache that data in memory. But that Cache() method needs us to *already* have a CloudCell? Why can't I just call ``CloudCell.Cache(myData)``? Why does it need to return ``local bool`` - why not just ``bool``?

It's difficult for me to explain away to people coming from frameworks like Hadoop or Spark the extra hoops that are needed to go through for fundamentals like "make this bit of data accessible across the cluster" which should be a one line operation to persist it + cache it, and returns a reference to that object.

```
let dataRef = handle.Cache(myData) // maybe an overload or operation to "force" eager caching?
let otherRef = handle.CacheFromStore("folder/subfolder/myblob") // read from existing blob
```

To me this should all be a single abstraction i.e. a way to get to some value across the cluster. The fact that it stores it in a blob and caches it in memory across the cluster as far as the end user is concerned doesn't matter - the only real abstraction they want to see is the reference to the blob of data after it's been sent to the cluster.

I'll try this weekend to do that typical use case analysis based on feedback I've seen so far.

**Comment by palladin** [comment]
@isaacabraham Regarding the CloudCell.Cache, I think that the following is what you want.
```
module CloudCell
let Cache (myData : 'T[]) = CloudCell.New(myData, cacheByDefault = true)
```
it stores it in a blob and caches it in memory in an on demand basis.

**Comment by eiriktsarpalis** [comment]
@isaacabraham I doubt 'better' abstractions will eliminate this class of problems. This is essentially an issue of continuation-based distributed computation. Even higher-level libraries like CloudFlow won't eliminate it. What can be done is implement under-the-hood optimizations in the runtime to reduce performance impact. We have certain ideas on this but they remain at speculative level atm. Your idea of issuing warnings if large values are caught travelling back-and-forth is a good one, and should not be difficult to implement.

Ultimately, I think we need to make users aware of these pitfalls by providing tutorials like the above.

**Comment by dsyme** [comment]
@isaacabraham @palladin Based on what I've learned I would have gone for CloudCell with caching turned on (I still don't really like that name and think it should maybe just b CloudBlob or CloudValue). CloudCacheable I've never used and I put it in the "advanced" bucket of things that shouldn't appear in the intellisense menus unless you opt-in to them.

Tutorials on a range of topics would be great.  To be honest I think we're on the right track. Keep reducing the number of "basic" abstractions down to a bare minimum. For example, it feels to me like Local, CloudChannel, CloudAtom, CloudSequence, CloudCacheable, the MBrace.Workflow commbinators and all the interface IFooBar abstractions are all "non-essential" for 75% of the uses of MBrace (for typical scale-out data-and-compute processing tasks).  I don't mind MBrace having these secondary abstractions but they need to be secondary, not first-day. Keep improving the tutorials and documentation for both basic and secondary abstractions.  Keep working on the orthogonal aspects such as diagnostics, error messages.   Keep trialling the system with fresh users and understanding what blocks them.

**Comment by isaacabraham** [comment]
@palladin: yes and no. A problem is that once you have something in the cache, I couldn't figure out how later to get a reference to the same item later on e.g. CloudCell.GetByKey("myData") or similar. Another problem is that there's no way of easily going from an existing blob straight into the cache.

@eiriktsarpalis: I understand where you're coming from, and perhaps you're right. I also know though that users often don't waste time with tutorials. They want to download a nuget package, hit intellisense and expect to get results quickly. This means really thinking about the first use experience and how to ensure that it's as easy as possible to get some positive results through perhaps a limited but easy-to-use API.

**Comment by eiriktsarpalis** [comment]
Ok, I think I'm starting to understand the source of this confusion. `CloudCell` as it stands is a value that has been persisted to a file in blob storage. The explicitness of this representation is evident by the fact that `CloudCell` carries a `.Path` property. As a secondary feature `CloudCell` is declared *cacheable* which is an effect external to its implementation, as happens with other primitives. In other words, they implement [this interface](https://github.com/mbraceproject/MBrace.Core/pull/53#issue-65673114).

I am starting to see now that you are looking for a different primitive. One whose underlying representation is an implementation detail not communicated in the API. What matters is replication and ubiquity. This can be achieved by `CloudCell` but this approach is not particularly emphasized.

Perhaps it's all an issue of naming. We could perhaps rename `CloudCell` to `FilePersistedValue` and CloudSequence to `FilePersistedValues` and have an `ICachedValue` ADT for cases where representation is not important.

**Comment by isaacabraham** [comment]
I thought CloudFile was the representation of something that is persisted in a blob?

I think you're on the right track with a cloud value I.e. Something that exists in the cluster. Even the notion of "caching" in the name is perhaps unnecessary.
________________________________

**Comment by eiriktsarpalis** [comment]
No, `CloudFile` is not typed. It's just a wrapper for a file path. It should be considered the cloud equivalent to `System.IO.FileInfo`, just like `CloudDirectory` is the cloud equivalent to `System.IO.DirectoryInfo`. I like to distinguish CloudFile from CloudCell in terms of the following correspondences:

CloudFile <--> System.IO.Stream <--> byte []
CloudCell<'T> <--> System.IO.Stream + deserializer for type 'T <--> object of type 'T

In other words, a CloudCell<> can be a CloudFile, but the opposite is not true. You also need to specify a encoding/deserializer for proper interpretation of the binary data. This is why `CloudCell.FromFile` has the type signature
``` CloudCell.FromFile : path:string * deserializer:(Stream -> 'T) -> Cloud<CloudCell<'T>> ```
and CloudCell.FromTextFile
``` CloudCell.FromTextFile : path:string * deserializer:(TextReader -> 'T) * encoding:Encoding -> Cloud<CloudCell<'T>>```

**Comment by dsyme** [comment]
Extending the description:

CloudFile <--> System.IO.Stream <--> byte[]/lines/text
CloudCell<'T> <--> System.IO.Stream + deserializer for type 'T + optional caching <--> object of type 'T
CloudSequence<'T> <--> System.IO.Stream + deserializer for type 'T <--> object of type 'T[]
CloudVector<'T> <--> System.IO.Stream + deserializer for type 'T + partitioning <--> object of type 'T[]
CloudCacheable <--> ??? <-> ???

leads me to think 

(a) CloudCacheable is mysterious, I don't know how it fits in
(b) there is conceptual redundancy between CloudSequence and CloudVector 
(c) I don't know where key-value stores fit into this picture, 

e.g. to ask 

* Why have CloudSequence when it is just a single-partition CloudVector?
* What is CloudCacheable?
* Why isn't CloudFile partitionable (and if I needed to partition a CloudFile what would I do?)

**Comment by eiriktsarpalis** [comment]
@dsyme to answer your questions
* Caching/CloudCacheable has been revised as of [this PR](https://github.com/mbraceproject/MBrace.Core/pull/53) but this is not present in the latest nuget release.
* CloudVector is an entity specific to MBrace.Flow that is built on top of CloudSequence. It encapsulates book-keeping functionality particular to this library. I'm currently working on [this branch](https://github.com/mbraceproject/MBrace.Core/tree/core-refactoring) that attempts to unify and extend the two concepts.
* The best alternative to key-value stores is CloudAtom, however it currently does not offer explicit control over keys.
* A CloudFile can be partitioned through the .Seek functionality of the reader stream. We already make use of this in `CloudFlow.ofTextFileByLine` for large files.

I should note that a more accurate description of CloudSequence in the context of the above diagrams is:

CloudSequence<'T> <--> System.IO.Stream + deserializer for type 'T <--> seq<'T>

The only reason why CloudSequence<'T> exists as an entity separate to CloudCell<'T []> is its capacity to deserialize large sequences of values on-demand, which may not fit in-memory in their entirety.

**Comment by dsyme** [comment]
Thanks for the links, you're working on the right things :)

OK, so:

 CloudFile <--> System.IO.Stream <--> byte[]/lines/text
 CloudCell<'T> <--> System.IO.Stream + deserializer for type 'T + optional caching <--> object of type 'T
 CloudSequence<'T> <--> System.IO.Stream + deserializer for type 'T <--> object of type seq<'T>
 CloudVector<'T> <--> System.IO.Stream + deserializer for type 'T + partitioning <--> object of type 'T[]
 CloudCacheable <--> computation generating 'T on demand + local caching  <~~~> Lazy<'T>

CloudCacheable feels like an odd name since CloudCell (and CloudSequence?) are also cacheable - I think this was the cause of confusion originally - CloudLazy may be better, with the docs saying it may be recomputed?

**Comment by dsyme** [comment]
| The only reason why CloudSequence<'T> exists as an entity separate to CloudCell<'T []> is its capacity to deserialize large sequences of values on-demand, which may not fit in-memory in their entirety.

I assume from this that CloudSequence and CloudCell do not have the same caching capabilities, and that it is not possible to cache CloudSequence.  Or should I add "optional caching" to CloudSequence above?

**Comment by eiriktsarpalis** [comment]
CloudCacheable is now an interface which both CloudCell and CloudSequence implement (CloudSequence is an instance of type ICloudCacheable<'T []>). This merely declares a potential for caching and does not encapsulate caching logic itself. So CloudSequence is perfectly capable for caching but this does not mean that you won't be getting OutOfMemory in certain large instances. This is why 'cacheByDefault' is set to false by default in both entities. This can be overridden by calling `.PopulateCache()` at any time; this will cache the payload in the current worker only.

The ability to cache CloudSequences is exercised heavily in the implementation of CloudVector in MBrace.Flow. Equally important is the ability to manually cache sequences in specific workers only: the partitions of a CloudVector do fit in memory however it is often the case that a whole CloudVector can never be materialized in a single machine.

**Comment by dsyme** [comment]
OK, thanks, so the summary is this (please feel free to edit to clarify)

| Cloud Type |  Underlying | Caching | Partitioned | Gives | Ongoing Work |
| --------:|:-----------:|:------------:|:---------:|:---------:|:-----:|
| `CloudFile` | `System.IO.Stream` | ?? | optional (undocumented) | byte[]/lines/text | |
| `CloudCell<'T>` | `System.IO.Stream` + deserializer for T | optional (off by default) | no | |
| `CloudSequence<'T>` | `System.IO.Stream` + deserializer for type 'T | optional (off by default) | no (see CloudVector) | `seq<'T>` | may be unified with `CloudVector` |
| `CloudVector<'T>` | `System.IO.Stream` + deserializer for type 'T | optional (on by default?) | yes |  `'T[]` | may be unified with `CloudSequence` |
| `CloudCacheable<'T>` | computation generating 'T on demand | yes (on by default?) | no | `Lazy<'T>` | |

Plus the various interfaces, the CloudAtom mutable abstraction and any future Key/Value store abstraction.  (Please add any other non-interface abstractions I've missed)

FWIW CloudCacheable still feels somewhat confusing as a name (ICloudCacheable is OK as an interface though. Newcomers looking for a construct for cached cell, vector and file data will naturally gravitate straight to that, when it will rarely be what they need. (I wonder if it can be removed in favour of a library function returning an ICloudCacheable)

**Comment by isaacabraham** [comment]
That's great, very useful to see. At a glance - from a consumer point of view - it looks like CloudVector and CloudFlow could be merged; CloudCell and CloudFile seem similar but not identical. Perhaps some easy combinators e.g. CloudCell.ofCloudFIle etc. etc. - and some simple static methods e.g. CloudCell.FromStore or similar.

I've been working on some use cases and some (crude) API extensions that might encourage some more discussion - will post in the next day or two.

**Comment by dsyme** [comment]
As an aside, the more we can make the whole thing feel like a unix-style file system or other-familiar-storage-system the better.  My impression is that the simplicity of HDFS is a huge part of Hadoop's success - the fact that people can just use ``/foo/bar/logs/2012-*-*/*`` (or something like that, I've forgotten the exact glob syntax) to reference an implicitly-partitioned collection files is a pretty remarkably cheap way of indexing the file system, and they can transfer all their Unix bash/sh/... knowledge of working with files to the HDFS distributed file system. (Speaking of which, an MBrace-to-Yarn mapping is one of the research projects I'm interested in this year)

@isaacabraham - assume you mean "CloudVector and CloudSequence"

Would it be reasonable to accept a collection of CloudCell into a CloudFlow? CloudCacheable? ICloudCacheable? 

Do we really expect CloudFiles to be partitioned using "Seek", practically speaking?  If so should a single CloudFile be an acceptable input into a CloudFlow? Would people actually use that?  I somehow expect not. (FWIW I tend to equate "partitioned" with "things I would use as source data for a CloudFlow").  Perhaps it would be better just to say these are not partitioned, but that they do support random access.

**Comment by eiriktsarpalis** [comment]
@dsyme thanks for taking the time to compile the table. I had an interesting brainstorming session with @palladin, we seem to have agreed on a set of refactorings in the store primitives. Without going into too much detail:
* Rename CloudCell<'T> to CloudFileValue<'T>. This should emphasise that it references a value that has been persisted to a single file in store. 
* Rename CloudSequence<'T> to CloudFileValues<'T>. This should emphasise that it references a sequence of values that has been persisted to a single file in store. 
* Add the abstract ICloudDictionary<'T> for distributed key/value storage.

We are also considering the following amendments:

* Remove CloudCacheable<'T>, in particular the 'lazy' workflow API. We discovered certain cases where pathological/naive use of the functionality would have detrimental effect to performance.
* Remove caching functionality from CloudCell/CloudSequence entirely. The CloudFlow library relies heavily on this capability se we need to proceed with caution if at all.
* Add an abstract ICloudCacheValue<'T> entity for value caching. This would work unlike CloudFileValue/CloudCell which carry concrete representations.
* Supposing that the idea in https://github.com/mbraceproject/MBrace.Core/issues/58#issuecomment-90067331 catches on, the purpose of CloudCell starts to become questionable. It might be reasonable to have it removed once this runtime optimisation is implemented.

I will update the table once details have been sorted out.

**Comment by dsyme** [comment]
OK, cool

FWIW I added a comment about the default of "on" for caching for CloudVector here: https://github.com/mbraceproject/MBrace.Core/commit/645ae37d2324f708ba1c63395c43332cee3f029a#commitcomment-10591132

The name "CloudFileValue(s)" names somehow don't quite resonate with me, at least not on first sight. Particularly "CloudFileValues" sounds like a .NET 1.x collection of "CloudFileValue".  

Perhaps "CloudFileAsValue" and "CloudFileAsSequence" though both are a bit long. 

Either way at least it's quite nice that all (?) persisted storage entities begin with "CloudFile".

**Comment by eiriktsarpalis** [comment]
I just created PR (#59) that addresses some of the discussed issues. Please review, I will update the table once it is merged.

**Comment by isaacabraham** [comment]
We're getting much closer to a fixed set of abstractions - can this be closed now I wonder (or replaced with another issue e.g. document abstractions on website).

**Comment by dsyme** [comment]
Yes - close it and replace by a "document abstractions" issue
